cpp/src/arrow/dataset/scanner_test.cc

```diff
@@ -163,6 +164,21 @@ struct MockScanTask {
       Future<std::shared_ptr<RecordBatch>>::Make();
 };
 
+// Wraps access to std::default_random_engine to ensure only one thread
+// at a time is using it
+class ConcurrentGen {
+ public:
+  explicit ConcurrentGen(std::default_random_engine* gen) : gen_(gen) {}
+  void With(std::function<void(std::default_random_engine*)> task) {
```

Comment:
really clueless nit: does `std::function` inhibit inlining of the lambda below?

Reply:
Yes, it does, regrettably.

Reply:
I suppose it doesn't really matter in this instance.

Reply:
Do you want me to change to a template parameter?  I was opting for readability here since it is unit test.

Reply:
No - I think it's fine, was more curious than anything else. (Especially since it's a test.)

Reply:
https://vittorioromeo.info/index/blog/passing_functions_to_functions.html is a pretty good summary of the cost.  I did some micro-benchmarks on this in the spring and got similar results.